testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/messaging/jms/context/injectedToRequestScoped/InjectedJMSContextToRequestScopedBeanTestCase.java

```diff
@@ -0,0 +1,88 @@
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.as.test.integration.messaging.jms.context.injectedToRequestScoped;
+
+import static org.junit.Assert.assertNull;
+
+import javax.inject.Inject;
+import javax.jms.JMSException;
+import javax.jms.Message;
+
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.junit.Arquillian;
+import org.jboss.as.arquillian.api.ServerSetup;
+import org.jboss.as.arquillian.api.ServerSetupTask;
+import org.jboss.as.arquillian.container.ManagementClient;
+import org.jboss.as.test.integration.common.jms.JMSOperations;
+import org.jboss.as.test.integration.common.jms.JMSOperationsProvider;
+import org.jboss.as.test.shared.TimeoutUtil;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+@RunWith(Arquillian.class)
+@ServerSetup({ InjectedJMSContextToRequestScopedBeanTestCase.JmsQueueSetup.class })
+public class InjectedJMSContextToRequestScopedBeanTestCase {
+
+    @Inject
+    private Alpha alpha;
+
+    static class JmsQueueSetup implements ServerSetupTask {
+
+        private JMSOperations jmsAdminOperations;
+
+        @Override
+        public void setup(ManagementClient managementClient, String containerId) throws Exception {
+            jmsAdminOperations = JMSOperationsProvider.getInstance(managementClient);
+            jmsAdminOperations.createJmsQueue("jmsContext/test", TestMessageListener.QUEUE_JNDI_NAME);
+        }
+
+        @Override
+        public void tearDown(ManagementClient managementClient, String containerId) throws Exception {
+            if (jmsAdminOperations != null) {
+                jmsAdminOperations.removeJmsQueue("jmsContext/test");
+                jmsAdminOperations.close();
+            }
+        }
+    }
+
+    @Deployment
+    public static JavaArchive createTestArchive() {
+        return ShrinkWrap.create(JavaArchive.class, "InjectedJMSContextToRequestScopedBeanTestCase.jar")
+                .addPackage(InjectedJMSContextToRequestScopedBeanTestCase.class.getPackage())
+                .addClass(TimeoutUtil.class)
+                .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml");
+    }
+
+    @Test
+    public void bothRequestScopedBeansDestroyed() throws JMSException, InterruptedException {
```

Comment:
What is this test actually doing?

You are sending a JMS message (from `alpha.ping()`) that is consumed by a MDB.
This MDB calls charlie and delta ping method (which does nothing or create a JMS message that is never sent anywhere).
You then wait 2 seconds to receives a message from the JMS queue but there will be none.
And then you check that the charlie and delta beans have properly been destroyed.

Is that correct?

Reply:
@jmesnil Hi, well it has already been some time ago. TBH it looks little bit weird to me as well, but you are basically right. It is about proper destruction of requestscoped beans. I think originally this resulted to the ConcurrentModificationException as described in the issue.